scripts/ensure-bundle-deps.ts

```diff
@@ -0,0 +1,146 @@
+  const { maxRetries = 12, dependencies } = options;
+
+  const processDotEnv = process.env as ProcessDotEnv;
+
+  // Did this build get invoked by a repo-level script, or a package-level script, and which script was it?
+  const isTopLevelBuild = path.basename(processDotEnv.INIT_CWD) === 'sentry-javascript';
+  const yarnScript = (JSON.parse(processDotEnv.npm_config_argv) as { original: string[] }).original[0];
+
+  // convert '@sentry/xyz` to `xyz`
+  const dependencyDirs = dependencies.map(npmPackageName => npmPackageName.split('/')[1]);
+
+  // The second half of the conditional tests if this script is being run by the original top-level command or a
+  // package-level command spawned by it.
+  const packagesDir = isTopLevelBuild && processDotEnv.INIT_CWD === process.cwd() ? 'packages' : '..';
+
+  if (checkForBundleDeps(packagesDir, dependencyDirs)) {
+    // We're good, nothing to do, the files we need are there
+    return;
+  }
+
+  // If we get here, the at least some of the dependencies are missing, but how we handle that depends on how we got
+  // here. There are six possibilities:
+  // - We ran `build` or `build:bundle` at the repo level
+  // - We ran `build` or `build:bundle` at the package level
+  // - We ran `build` or `build:bundle` at the repo level and lerna then ran `build:bundle` at the package level. (We
+  //   shouldn't ever land here under this scenario - the top-level build should already have handled any missing
+  //   dependencies - but it's helpful to consider all the possibilities.)
+  //
+  // In the first version of the first scenario (repo-level `build` -> repo-level `build:bundle`), all we have to do is
+  // wait, because other parts of `build` are creating them as this check is being done. (Waiting 5 or 10 or even 15
+  // seconds to start running `build:bundle` in parallel is better than pushing it to the second half of `build`,
+  // because `build:bundle` is the slowest part of the build and therefore the one we most want to parallelize with
+  // other slow parts, like `build:types`.)
+  //
+  // In all other scenarios, if the dependencies are missing, we have to build them ourselves - with `build:bundle` at
+  // either level, we're the only thing happening (so no one's going to do it for us), and with package-level `build`,
+  // types and npm assets are being built simultaneously, but only for the package being bundled, not for its
+  // dependencies. Either way, it's on us to fix the problem.
+  //
+  // TODO: This actually *doesn't* work for package-level `build`, not because of a flaw in this logic, but because
+  // `build:rollup` has similar dependency needs (it needs types rather than npm builds). We should do something like
+  // this for that at some point.
+
+  if (isTopLevelBuild && yarnScript === 'build') {
+    let retries = 0;
+
+    console.log('\nSearching for bundle dependencies...');
+
+    while (retries < maxRetries && !checkForBundleDeps(packagesDir, dependencyDirs)) {
```

Comment:
nit: This should really be a helper func which just retries a generic func a certain num of times.

Reply:
Disagree. I feel like there's enough bespoke stuff here that refactoring it into a generic retry-er would take some effort, and AFAIK we don't have a general need for such a thing at the moment. I'd rather put off that work until it's actually needed.